### ISSUE TITLE
would it be possible to remove dependancy on pyreadline ?

### ISSUE THREAD

**Issue by stonebig** [uqfoundation/dill#181]
indeed:
- pyreadline for Windows is a problem.
- the jupyter/ipython team replaced it by python-prompt, so there seems to be a way.

**Comment by mmckerns** [comment]
Yes, I'm looking into it since #171.  I'll do this as soon as I have the time.  Feel free to submit a pull request if you can get to it before me.  `pyreadline` is only used to read the buffer for interactively defined objects, so it's only in the `dill` code in one place I believe, and it is something that `python-prompt` can handle.

**Comment by stonebig** [comment]
now I have the problem because of CVXPY

pyreadline>=1.7.1 (from dill>=0.2.5->multiprocess->cvxpy

**Comment by mmckerns** [comment]
@stonebig: `dill` can't do what `jupyter` has done, as the reason `python-prompt` works for them is that they have their own console they can modify -- so basically they append every line that is entered into their `jupyter` console to the `python_prompt` history instance.  If you are using the python interpreter (i.e. not `IPython`/`jupyter`), then that trick doesn't work.  `pyreadline` is the only good general solution on windows that I can find.  Yeah, it's end of life, but I don't foresee it going away at any time, and It's pure python should not be difficult to install.

What is the problem with `pyreadline` for you on windows?  I don't understand the issue. For me, it's easy to install.  Please elaborate.

I can very easily put in a `try` / 'except' around the one place that `dill` does a `import readline`… and then could make the `readline` install optional in `setup.py`.  However, I'll need a good reason to do that.  Can you help convince me that's a good idea?

**Comment by mmckerns** [comment]
There is this PR which has what is needed: https://github.com/jonathanslenders/python-prompt-toolkit/pull/99/files, however, it's not matching the current interface to `prompt_toolkit.interface.CommandLineInterface`, so it would need work.

Also, three's another module mentioned: https://bitbucket.org/pypy/pyrepl/ ...so I'll look at it too.

**Comment by mmckerns** [comment]
Regardless, @stonebig: what's the reason for replacing `pyreadline`?  is it just that it is not being maintained going forward?

**Comment by stonebig** [comment]
it's classified as "(unstable, deprecated)" per Windows community.

Jupyter project got rid of it for this reason.

**Comment by mmckerns** [comment]
`pyreadline` was developed by the `jupyter` team, if I'm not mistaken.  They have a suitable replacement in `python_prompt` as they have their own console.  `dill` doesn't have a console, so until the above-referenced PR gets merged, or someone mods the code in that PR so it works to emulate `readline`… then there is no foreseeable alternate solution.  Using unmaintained code is typically not an issue until there is some community-wide change that it doesn't keep up with, especially when the dependency is only for a 2-liner as it is in `dill`.

So, if your only beef with `pyreadline` is that it's unmaintained, then it's not a pressing issue for `dill` at this time.  However, if you have something else that makes it a pressing issue (i.e. there's an existing *problem* using or installing `pyreadline`), then please let me know and more immediate action will be taken.

**Comment by mmckerns** [comment]
@stonebig: Can you help me understand how `pyreadline` is a "problem"?  What I mean by a problem, is: why exactly is it causing you or the package(s) you maintain an issue?  As the combination of `pyreadline` and `dill` work for me on all versions of python on windows that I have tested, and I don't see what the issue might be (aside from depending on a unmaintained package) -- I'm going to close this issue. Please feel free to reopen it if you can better explain what the problem you are experiencing is.  As I said, I'd like to replace `pyreadline`, but I don't see a suitable replacement for it.  Also, please feel free to reopen this if you find one.

**Comment by ankostis** [comment]
Readline in PY3.6 raises [`AttributeError: module 'readline' has no attribute 'redisplay'`](/pyreadline/pyreadline/issues/49) if you press `[Tab]` inside a python interpreter.
Uninstalling it it is not an option due to *dill*.

So I believe this should re-open.

**Comment by ankostis** [comment]
Maybe turn it into an *setuptools* [`extras_require`](http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) and leave those interested to install it?

**Comment by ankostis** [comment]
FYI, untill Untill pyreadline/pyreadline#56 gets merged, you may workaround this by installing this patched `pyreadline` brach: 

    pip install git+https://github.com/ankostis/pyreadline@redisplay

**Comment by CutiePi** [comment]
@ankostis Worked like a charmed on Python 3.6.1 (v3.6.1:69c0db5, Mar 21 2017, 17:54:52) [MSC v.1900 32 bit (Intel)] on win32

## Thanks bud

**Comment by stonebig** [comment]
now, i don't remember why i needed it...

**Comment by mmckerns** [comment]
@stonebig: ok... so that is taking matters into your own hands.  Cool.  The problem for me is that making `readline` an optional dependency is not feasible, it's too integral of a feature to be able to read from the buffer.  However, your fork does provide for a potential solution going forward.

What do you think of either (1) releasing the fork as a package on pypi, or (2) `dill` using a tarball of your fork on GitHub as a dependency?  Either way, `dill` could depend on that instead of `readline`, at least for the short term.  See: http://python-packaging.readthedocs.io/en/latest/dependencies.html#packages-not-on-pypi. 

I think the better option is (1), because, I believe that (2) will cause some slight issues as noted here: https://stackoverflow.com/questions/29170630/how-to-allow-unverified-packages-in-requirements-txt

So, for me, an easy fix for the short term is for you to release your fork as a package, and `dill` can depend on it if `readline` is not found (or vice-versa).  How does that sound?

**Comment by mmckerns** [comment]
I will cut a new release as soon as this can be resolved.  My plan was to release this week, but I might push it back as this would be a major change.

**Comment by stonebig** [comment]
sorry was lost for open source last 2 Weeks. will retry to integrate pyreadline again soon. I don't like forks

**Comment by mmckerns** [comment]
I don't like forks either, but it would hopefully be temporary -- and a fork is better than nothing.  Let me know what I can do to help get your PR accepted.  It doesn't look like there's a lot of movement from `pyreadline`.

**Comment by stonebig** [comment]
I re-included pyreadline for the time being.

**Comment by mmckerns** [comment]
@stonebig: so I'm going to cut a patch release by the end of the week.  How should I point `dill` to your code?  

Would you consider releasing it so it can be installed with `pip install your_readline`? or would you rather I point `dill` to your code on GitHub (e.g. as in the link above for "packages not on pypi")?  I'd rather point to a released package on pypi...

**Comment by stonebig** [comment]
I can compile a "dill" from source, I suppose, just point me to the tag that would better fit me.

**Comment by mmckerns** [comment]
Wait.  I think there's some misunderstanding here.  What I'm saying is that since your patched branch of `pyreadline` fixes the issue that's a blocker for you -- and `dill` needs `pyreadline` on Windows -- maybe it makes sense to point `dill` to your branch instead of `pyreadline`, at least until pyreadline/pyreadline#56 is merged.   I think that's possible, but not desirable. I'd rather do is point to a released package over an unreleased package.   I'm not telling you to fork `dill`.  I'm saying that I can change the requirements of `dill` to use your branch over `pyreadline` -- and it would help if you cut a release I could point to.  Does that make sense?
